fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/LAONetworkManager.java

```diff
@@ -0,0 +1,194 @@
+        .subscribeOn(schedulerProvider.newThread())
+        .subscribe(this::handleGenericMessage);
+  }
+
+  @Override
+  public Completable catchup(String channel) {
+    Log.d(TAG, "sending a catchup to the channel " + channel);
+    Catchup catchup = new Catchup(channel, requestCounter.incrementAndGet());
+    return request(catchup)
+        .map(ResultMessages.class::cast)
+        .map(ResultMessages::getMessages)
+        .doOnSuccess(messages -> Log.d(TAG, "Catchup on " + channel + " retrieved : " + messages))
+        .doOnSuccess(messages -> handleMessages(messages, channel))
+        .ignoreElement();
+  }
+
+  @Override
+  public Completable publish(KeyPair keyPair, String channel, Data data) {
+    return publish(channel, new MessageGeneral(keyPair, data, gson));
+  }
+
+  @Override
+  public Completable publish(String channel, MessageGeneral msg) {
+    Log.d(TAG, "sending a publish " + msg.getData().getClass() + " to the channel " + channel);
+    Publish publish = new Publish(channel, requestCounter.incrementAndGet(), msg);
+    return request(publish).ignoreElement();
+  }
+
+  @Override
+  public Completable subscribe(String channel) {
+    Log.d(TAG, "sending a subscribe on the channel " + channel);
+    Subscribe subscribe = new Subscribe(channel, requestCounter.incrementAndGet());
+    return request(subscribe)
+        // This is used when reconnecting after a lost connection
+        .doOnSuccess(answer -> subscribedChannels.add(channel))
+        .doAfterSuccess(answer -> catchup(channel))
+        .ignoreElement();
+  }
+
+  @Override
+  public Completable unsubscribe(String channel) {
+    Log.d(TAG, "sending an unsubscribe on the channel " + channel);
+    Unsubscribe unsubscribe = new Unsubscribe(channel, requestCounter.incrementAndGet());
+    return request(unsubscribe)
+        // This is used when reconnecting after a lost connection
+        .doOnSuccess(answer -> subscribedChannels.remove(channel))
+        .ignoreElement();
+  }
+
+  private void handleGenericMessage(GenericMessage genericMessage) {
+    Log.d(TAG, "handling generic msg : " + genericMessage);
+    if (genericMessage instanceof Broadcast) {
+      // We've a Broadcast
+      Broadcast broadcast = (Broadcast) genericMessage;
+      try {
+        messageHandler.handleMessage(
+            repository, this, broadcast.getChannel(), broadcast.getMessage());
+      } catch (DataHandlingException e) {
+        Log.e(TAG, "Error while handling received message", e);
+        unprocessed.onNext(genericMessage);
+      }
+    }
+  }
+
+  private void handleMessages(List<MessageGeneral> messages, String channel) {
+    for (MessageGeneral msg : messages) {
+      try {
+        messageHandler.handleMessage(repository, this, channel, msg);
+      } catch (DataHandlingException e) {
+        Log.e(TAG, "Error while handling received catchup message", e);
+      }
+    }
+  }
+
+  private Single<Answer> request(Query query) {
```

Comment:
this function could benefit from a few inline comments explaining the intention behind the magic 😄